File: GFG/Tree/balancedtree.cpp
#include<iostream>
#include"../ds.h"
using namespace std;

int height(TNode<int>* root){
    return (root==NULL) ? 0 : max(height(root->left),height(root->right))+1;
}
bool isbal(TNode<int>* root){
    if(root==NULL){
        return true;
    }
    int lh=height(root->left);
    int rh=height(root->right);
    return (abs(lh-rh)<=1 && isbal(root->left) && isbal(root->right));

}

int main(){
    TNode<int>* root=new TNode(40);
    root->left=new TNode(30);
    root->left->left=new TNode(25);
    root->left->left->left=new TNode(15);
    root->left->right=new TNode(35);
    root->left->left->right=new TNode(28);
    root->right=new TNode(50);
    root->right->left=new TNode(45);
    root->right->right=new TNode(60);
    root->right->right->right=new TNode(70);
    root->right->right->left=new TNode(55);
    cout<<isbal(root);
    return 0;
}

File: GFG/Tree/balancedtree.py
import ds

def height(root):
    return 0 if (root is None) else max(height(root.left),height(root.right))+1

def isbal(root):
    if(root is None):
        return True
    lh=height(root.left)
    rh=height(root.right)
    return( abs(lh-rh)<=1 and isbal(root.left) and isbal(root.right))

root=ds.TNode(40)
root.left=ds.TNode(30)
root.left.left=ds.TNode(25)
root.left.left.left=ds.TNode(15)
root.left.right=ds.TNode(35)
root.left.left.right=ds.TNode(28)
root.right=ds.TNode(50)
root.right.left=ds.TNode(45)
root.right.right=ds.TNode(60)
root.right.right.right=ds.TNode(70)
root.right.right.left=ds.TNode(55)
print(isbal(root))